core/matrix/fbcsr.cpp

```diff
@@ -0,0 +1,472 @@
+void Fbcsr<ValueType, IndexType>::move_to(
+    SparsityCsr<ValueType, IndexType> *const result)
+{
+    this->convert_to(result);
+}
+
+
+/* Within blocks, the storage order is row-major.
+ * Currently, this implementation is sequential and has complexity O(n log n)
+ * assuming nnz = O(n).
+ * Can this be changed to a parallel O(n) implementation?
+ */
+template <typename ValueType, typename IndexType>
+void Fbcsr<ValueType, IndexType>::read(const mat_data &data)
+{
+    if (data.nonzeros.size() > std::numeric_limits<index_type>::max())
+        throw std::range_error(std::string("file: ") + __FILE__ + ":" +
+                               std::to_string(__LINE__) +
+                               ": List of nonzeros is too big!");
+
+    const index_type nnz = static_cast<index_type>(data.nonzeros.size());
+
+    const int bs = this->bs_;
+
+    using Blk_t = blockutils::DenseBlock<value_type>;
+
+    struct FbEntry {
+        index_type block_row;
+        index_type block_column;
+    };
+
+    struct FbLess {
+        bool operator()(const FbEntry &a, const FbEntry &b) const
+        {
+            if (a.block_row != b.block_row)
+                return a.block_row < b.block_row;
+            else
+                return a.block_column < b.block_column;
+        }
+    };
+
+    auto create_block_map = [nnz, bs](const mat_data &data) {
+        std::map<FbEntry, Blk_t, FbLess> blocks;
+        for (index_type inz = 0; inz < nnz; inz++) {
+            const index_type row = data.nonzeros[inz].row;
+            const index_type col = data.nonzeros[inz].column;
+            const value_type val = data.nonzeros[inz].value;
+
+            const int localrow = static_cast<int>(row % bs);
+            const int localcol = static_cast<int>(col % bs);
+            const index_type blockrow = row / bs;
+            const index_type blockcol = col / bs;
+
+            Blk_t &nnzblk = blocks[{blockrow, blockcol}];
+            if (nnzblk.size() == 0) {
+                nnzblk.resize(bs, bs);
+                nnzblk.zero();
+                nnzblk(localrow, localcol) = val;
+            } else {
+                if (nnzblk(localrow, localcol) != gko::zero<value_type>())
+                    throw Error(__FILE__, __LINE__,
+                                "Error: re-visited the same non-zero!");
+                nnzblk(localrow, localcol) = val;
+            }
+        }
+        return blocks;
+    };
+
+    const std::map<FbEntry, Blk_t, FbLess> blocks = create_block_map(data);
+
+    auto tmp = Fbcsr::create(this->get_executor()->get_master(), data.size,
+                             blocks.size() * bs * bs, bs);
+
+    tmp->row_ptrs_.get_data()[0] = 0;
+    index_type cur_brow = 0, cur_bnz = 0,
+               cur_bcol = blocks.begin()->first.block_column;
+    const index_type num_brows = data.size[0] / bs;
+
+    blockutils::DenseBlocksView<value_type, index_type> values(
+        tmp->values_.get_data(), bs, bs);
```

Comment:
I use the `DenseBlocksView` to deal easily with 2D indexing within blocks. I could use an `Array::view` inside, but it just seems simpler this way. I thought a `value_type *const` should be safe enough since this class does no memory management.